components/server/src/server.rs

```diff
@@ -153,12 +155,37 @@ fn run_impl<CER: ConfiguredRaftEngine, F: KvFormat>(config: TikvConfig) {
     tikv.run_status_server();
     tikv.core.init_quota_tuning_task(tikv.quota_limiter.clone());
 
-    signal_handler::wait_for_signal(
-        Some(tikv.engines.take().unwrap().engines),
-        tikv.kv_statistics.clone(),
-        tikv.raft_statistics.clone(),
-    );
+    // Build a background worker for handling signals.
+    let handler = {
+        let engines = tikv.engines.take().unwrap().engines;
+        let kv_statistics = tikv.kv_statistics.clone();
+        let raft_statistics = tikv.raft_statistics.clone();
+        std::thread::spawn(move || {
+            signal_handler::wait_for_signal(
+                Some(engines),
+                kv_statistics,
+                raft_statistics,
+                Some(service_event_tx),
+            )
+        })
+    };
+    loop {
+        if let Ok(service_event) = service_event_rx.recv() {
+            match service_event {
+                ServiceEvent::PauseGrpc => {
+                    tikv.pause();
+                }
```

Comment:
If we want abort requests ASAP, how about let them select a abort signal?

I think it's less error-prone, as it does not need to restart gRPC server.

For all gRPC requests,
``` rust
let (abort, notify) = tokio::sync::broadcast::channel(1);

if server is slow {
    drop(notify);
}

fn kv_requests(ctx, req) {
    ctx.spawn(async {
        select! {
          handle(req).await => { .. }
          abort.await => { .. }
        }
    })
}
```

Reply:
We may try this approach later. For now let's stick to restarting gRPC server as the short term solution. 
The reason is that the approach above is essentially is "api level grpc restart",  it needs to handle not only current requests, but also prevent new requests reaching to this TiKV node. Also if we need to add a new error code, then it would need tikv-client's code change.  It's all do-able, but it requires more time to implement and test. 

As discussed, we will use restart grpc server as the first short term solution. And we can explore the suggested solution later once we have a workable baseline.